### ISSUE TITLE
XML Bird and GXml

### ISSUE THREAD

**Issue by esodan** [johanmattssonm/xmlbird#3]
I'm maintaining GXml at:

https://git.gnome.org/browse/gxml

It is using libxml2 backend, but I'm searching for a different engine to parse XML files.

GXml is written in Vala and have a set of interfaces to access XML files, with two implementations one to read/write XML documents based on libxml2 Xml.Node structures (slow) and one just to write (no parse jet but 10 time faster) using libxml2 Xml.TextWriter.

GXml use this interfaces to provide GObject to XML serialization framework, with support of Gee collections and custom implementations.

I'm planning to implement GXml's interfaces in a set of classes, to help to serialize/deserialize directly from/to GObject properties, without parse first to/from an XML tree, improving speed.

I would like to now if you are are interested to work together. Like implementing GXml interfaces using your classes to parse XML documents, compare on speed and check if we can benefit each other.

May I can help you to add Autotools support to your project, this will help you to share your work with other projects, like GXml.

**Comment by johanmattssonm** [comment]
Autotools and integration with GXml sounds like two very good steps to take. I will probably keep the doit framework side by side with Autotools for a while because Autotools often fails on windows but a contribution would be highly appreciated.

There are several things that needs improvement in libxmlbird. The time complexity for parsing nested tags is not good at all and I think a bit of work needs to be done in order to comply with the XML specification. I am very interested in collaborating with you around creating a fast XML parser.

Is your plan to use libxmlbird just for parsing and manage the DOM tree in GXml?

**Comment by esodan** [comment]
I have GXml building very confident on windows, with GXml Autotools configuration, then I can port this to your project, including support for Windows and documentation.

My work will be focused on GXml Serialization framework, but any other XML support will be on the radar too. In the future may be support for XSD and other technologies will be fine for GXml.

libxmlbird is a good step forward for XML parsing, but not just for DOM, but any other XML struct, because GXml have a DOM API but depends on libxml2 Xml.Node parsing, I would like to change that by porting DOM API to interfaces then implement them with the actual now, a few, legacy implementation in order to change used engines.

libxml2, have good support from standard and provides a Xml.TextReader objects, I would like to use to parse XML documents to a Gee collection based tree implementing GXml interfaces (this class is Tw prefixed - TwNode, for example).

TwDocument is a class implementing GXml.Document interface and use libxml2's TextWriter to write down XML documents in its tree, it is very fast compared with Xml.Node tree from libxml2.

May you can help me on TwDocument to implement a Xml.TextReader based XML parser, think about libxml2 long history and standard compliant on XML.

libxml2 have parallel projects to support other standards but because it is writing in plain C, helps it on speed but lacks most of object oriented features required to implement complex things like XSD.

GXml, as it is written in Vala, provides a good Object Oriented XML environment to develop complex implementations, this is the base for GObject Serialization to XML where the user can define GObject classes derived from, for example, GXml.SerializableObjectModel to write/read to/from XML files its GObject's public properties.

I've helped GtkFlow to be ported to Autotools too and works great on Windows, just as a side note :-)

**Comment by johanmattssonm** [comment]
A stream based XML reader would certainly solve the performance issue. I will write a tiny XML reader implementation but I am not sure when it eventually will become ready to use.

**Comment by esodan** [comment]
I think we should not reinvent the weel.

Today I started an upgrade and found security updates for libxml2 for
malformed or specific conditions in XmlReader, this could be the case for
xmlbird or any other young piece of software.

libxml2 is C but its reader could be  used to parse and fill GObject's
properties using its Vala bindings. This is the case for GXml.TwDocument in
simple jet powerful function that iterates in Gee collections with GObjects
properties. Check it out at:

https://git.gnome.org/browse/gxml/tree/gxml/TwDocument.vala

For write_document() method, line 124

May you, and certainly GXml too, can use libxml2 XmlReader to stream tags
and its properties, including other types like annotations, comments,
namespace, and others, with a well proven security.

**Comment by esodan** [comment]
Can we take over this issue again?
GXml has evolved up to have an interface for XML parsers, so may is a matter to implement it using XMLBird as a new backend.
I really love to see a parser that can:
* Read a tag and its properties, but skip all its content, by returning it to create a threat to parse internal tags
* Once the new thread is created for the previous tag, jump to seek a new Tag and repeat last step
* At the end we will have a set of tasks, running in parallel in threads (a Thread pool)

Hope these steps can help to improve XML tree parsing.

Hope you are open to work on this. I can add Meson to your project, if so I can open a new issue.

**Comment by johanmattssonm** [comment]
Hi. I am probably not the right person to write this code.

**Comment by esodan** [comment]
Hi. GXml now has support for the future described at https://github.com/johanmattssonm/xmlbird/issues/3#issuecomment-475033234 it is called StreamReader

I had to write a custom basic XML parser to do so, you can check it out at: https://gitlab.gnome.org/GNOME/gxml
